datalad/support/gitrepo.py

```diff
@@ -2355,17 +2358,67 @@ def get_submodules_(self, paths=None):
             return
 
         modinfo = self._parse_gitmodules()
-        for path, props in self.get_content_info(
-                paths=paths,
-                ref=None,
-                untracked='no').items():
-            if props.get('type', None) != 'dataset':
+        if not modinfo:
+            # we exit early, if there is nothing on record (even though
+            # a .gitmodules file exists).
+            # without this conditional exit, we would be able to discover
+            # subprojects even when they are not recorded in .gitmodules,
+            # but at the cost of running an unconstrained ls-files call
+            # below
+            return
+
+        # taken from 3.9's pathlib, can be removed once the minimally
+        # supported python version
+        def is_relative_to(self, *other):
+            """Return True if the path is relative to another path or False.
+            """
+            try:
+                self.relative_to(*other)
+                return True
+            except ValueError:
+                return False
+
+        if paths:
+            # ease comparison
+            paths = [self.pathobj / p for p in paths]
+            # contrain the report by the given paths
+            modinfo = {
+                # modpath is absolute
+                modpath: modprobs
+                for modpath, modprobs in modinfo.items()
+                # is_relative_to() also match equal paths
+                if any(is_relative_to(modpath, p) for p in paths)
```

Comment:
Mutating input variables often makes it harder to debug that code later since causes confusion (docstring says `paths` is relative -- how they are absolute?), so I would have preferred:

```suggestion
            abspaths = [self.pathobj / p for p in paths]
            # contrain the report by the given paths
            modinfo = {
                # modpath is absolute
                modpath: modprobs
                for modpath, modprobs in modinfo.items()
                # is_relative_to() also match equal paths
                if any(is_relative_to(modpath, p) for p in abspaths)
```

Reply:
note that this is potentially `O(N^2)` operation (becomes expensive) if paths provided for all submodules (e.g smth like `subdatasets sub-*` where each sub- is a subdataset). It reminded me about [get_parent_paths](https://github.com/datalad/datalad/blob/HEAD/datalad/support/path.py#L119) solving similar problem -- please see if it could be used here as well.  If not -- we should be able to come up with at least `O(N*log(N))` solution here

Reply:
didn't check if that is due to this particular code path but here is comparative timing of master (<.3sec) and this branch (>3 sec):

```shell
(git)lena:~datalad/datalad-master[master]git
$> datalad --version; (builtin cd /home/yoh/datalad/hcp-openaccess; time python -c "from datalad.support.gitrepo import *; from glob import glob; print('\n'.join(map(str,GitRepo('.').get_submodules_(glob('HCP1200/*')))))" | wc -l; )
datalad 0.17.2+102.g2c0073395
1113
python -c   0.26s user 0.04s system 107% cpu 0.277 total
wc -l  0.00s user 0.00s system 0% cpu 0.277 total

$> git co -
Switched to branch 'bf-6940'
Your branch is up to date with 'gh-mih/bf-6940'.

$> datalad --version; (builtin cd /home/yoh/datalad/hcp-openaccess; time python -c "from datalad.support.gitrepo import *; from glob import glob; print('\n'.join(map(str,GitRepo('.').get_submodules_(glob('HCP1200/*')))))" | wc -l; )
datalad 0.17.2+101.gb36148675
1113
python -c   3.78s user 0.02s system 100% cpu 3.777 total
wc -l  0.00s user 0.00s system 0% cpu 3.777 total

```

Reply:
I can replicate this. I have not looked into the reasons, but I will state upfront that "give me all subdatasets" is a more common use case than "give my all subdatasets for the paths of all subdatasets". Even if the slow-down for this use case is unavoidable it does not outweigh the substantial speed-up for the common case.

Reply:
I can confirm that applying the path-constraint is the source of the slowness. I have no idea how to fix this in a reasonable time frame. Please advice on how to proceed.

The remaining aspect in this conversation is

> docstring says paths is relative -- how they are absolute?

They are not absolute, as the docs say. But they have to become absolute to enable the comparison with the paths reported by `_parse_gitmodules()`.

The path conversion has no significant impact on the timing.

Reply:
>  I have no idea how to fix this in a reasonable time frame. Please advice on how to proceed.

I have advised in 

> It reminded me about [get_parent_paths](https://github.com/datalad/datalad/blob/HEAD/datalad/support/path.py#L119) solving similar problem -- please see if it could be used here as well.

I will check in detail if it could be used here now. Otherwise can come up with some other O(N*log(N)) strategies similar to annotate paths we used to do to assign paths into subdatasets paths (here it is assign submodule paths into `paths` -- kinda the same, and we probably already do that with some other functions as well...)

> The remaining aspect in this conversation is
> 
> > docstring says paths is relative -- how they are absolute?
> 
> They are not absolute, as the docs say. But they have to become absolute to enable the comparison with the paths reported by `_parse_gitmodules()`.

So, without the assignment to the dedicated `absname` if I end up at any line after you make them absolute within `paths` - the **are not absolute**, thus contradicting what the docs say and my expectation. that was the only concern. I have no concern on timing impact from such conversion.

Reply:
Ok, so it read from your statement that this PR is blocked until someone investigated whether this edge case can be made faster to some degree.

Reply:
here is a patch which seems to work as desired 
```patch
diff --git a/datalad/support/gitrepo.py b/datalad/support/gitrepo.py
index a437fc5a2..0776f5805 100644
--- a/datalad/support/gitrepo.py
+++ b/datalad/support/gitrepo.py
@@ -59,6 +59,7 @@ from datalad.dataset.gitrepo import (
     path_based_str_repr,
 )
 from datalad.log import log_progress
+from datalad.support.path import get_parent_paths
 from datalad.support.due import (
     Doi,
     due,
@@ -2367,32 +2368,14 @@ class GitRepo(CoreGitRepo):
             # below
             return
 
-        # taken from 3.9's pathlib, can be removed once the minimally
-        # supported python version
-        def is_relative_to(self, *other):
-            """Return True if the path is relative to another path or False.
-            """
-            try:
-                self.relative_to(*other)
-                return True
-            except ValueError:
-                return False
-
+        mod_relpaths = [str(m.relative_to(self.pathobj)) for m in modinfo]
         if paths:
-            # ease comparison
-            paths = [self.pathobj / p for p in paths]
-            # contrain the report by the given paths
-            modinfo = {
-                # modpath is absolute
-                modpath: modprobs
-                for modpath, modprobs in modinfo.items()
-                # is_relative_to() also match equal paths
-                if any(is_relative_to(modpath, p) for p in paths)
-            }
+            # constrain the report by the given paths
+            mod_relpaths = get_parent_paths(paths, mod_relpaths, only_with_parents=True)
         for r in self.call_git_items_(
             ['ls-files', '--stage', '-z'],
             sep='\0',
-            files=[str(p.relative_to(self.pathobj)) for p in modinfo.keys()],
+            files=mod_relpaths,
             read_only=True,
             keep_ends=True,
         ):
```
and brings it down to 

```shell
$> datalad --version; (builtin cd /home/yoh/datalad/hcp-openaccess; time python -c "from datalad.support.gitrepo import *; from glob import glob; print('\n'.join(map(str,GitRepo('.').get_submodules_(glob('HCP1200/*')))))" | wc -l ; )
datalad 0.17.2+102.g8b49a71d7.dirty
1113
python -c   0.18s user 0.03s system 110% cpu 0.194 total
wc -l  0.00s user 0.00s system 0% cpu 0.194 total
```

so as well faster than current master and removing the need for borrowed `is_relative` [I questioned above](https://github.com/datalad/datalad/pull/6942/files#r947882244).  Known annotated tests for get_submodules pass

```
$> DATALAD_TESTS_SSH=1 python -m pytest -s -v -k get_submodule datalad
======================================================================= test session starts =======================================================================
platform linux -- Python 3.10.5, pytest-7.1.2, pluggy-1.0.0 -- /home/yoh/proj/datalad/datalad-master/venvs/dev3/bin/python
cachedir: .pytest_cache
rootdir: /home/yoh/proj/datalad/datalad-master, configfile: tox.ini
plugins: fail-slow-0.2.0, cov-3.0.0
collected 1304 items / 1301 deselected / 3 selected                                                                                                               

datalad/support/tests/test_annexrepo.py::test_AnnexRepo_get_submodules SKIPPED (TODO)
datalad/support/tests/test_gitrepo.py::test_GitRepo_get_submodules PASSED
datalad/support/tests/test_gitrepo.py::test_get_submodules_parent_on_unborn_branch PASSED
```

but I feel that we might benefit from more tests here!
```

Reply:
I think this would not work on Windows. At least the docs of that function say that it needs posix paths. So it seems an explicit conversion to posix is necessary, and in a way that is consistent with what git does for the report from ls-files.

Reply:
NB note that appveyor tests on windows pass. I feel that there might be a need for a dedicated test here for any type of fix.

indeed -- there is that docstring!  and more confusing:
- at https://github.com/datalad/datalad/blob/master/datalad/support/path.py#L183 we use hardcoded `'/'`
- in `_get_parent_paths_check` at https://github.com/datalad/datalad/blob/master/datalad/support/path.py#L202 we use `os.sep`

and both cases are on `path` of `paths`, and both are added within the same commit where that function was added 5fefda4370e4beac5e5687846e801c9f9f0dadf1 .

Looking at the code, I do not see any "POSIX-specifics" of any kind, so I think that the comment and hardcoded use of `/` *could go* and make POSIX assumption optional (so use on git output could still be used).  I will propose a PR for that because I do feel that currently that function could indeed be buggy. 

proposed patch could do what we do in the other use of that function within `get_content_info` -- `paths = map(ut.PurePosixPath,  paths)` first to ensure that they are POSIX, after all we are to compare to modules paths which should be POSIX right?

Reply:
submitted https://github.com/datalad/datalad/pull/6963 . While looking at the code realized that above I am pretty much suggesting what `get_content_info` was doing when called from `get_submodules_()` so to some effect we are just moving that logic out. Indeed would need aforementioned `map(ut.PurePosixPath, paths)` to be added to the patch but otherwise -- should be fine.